internal/backend/grpc/router/router.go

```diff
@@ -173,10 +173,28 @@ func (r *Router) Director(ctx context.Context, fullMethodName string) (proxy.Mod
 func (r *Router) getTalosBackend(ctx context.Context, md metadata.MD) ([]proxy.Backend, error) {
 	clusterName := getClusterName(md)
 
+	nodeID := getNodeID(md)
+
 	id := fmt.Sprintf("cluster-%s", clusterName)
 
 	if clusterName == "" {
-		id = fmt.Sprintf("machine-%s", getNodeID(md))
+		machineSetNode, err := safe.ReaderGetByID[*omni.MachineSetNode](actor.MarkContextAsInternalActor(ctx), r.cosiState, nodeID)
+		if err != nil && !state.IsNotFoundError(err) {
+			return nil, fmt.Errorf("failed to resolve machine cluster relation %w", err)
+		}
+
+		if machineSetNode != nil {
+			var ok bool
+
+			clusterName, ok = machineSetNode.Metadata().Labels().Get(omni.LabelCluster)
+			if !ok {
+				return nil, fmt.Errorf("failed to resolve machine cluster relation: no cluster name in the machine set node %q", id)
+			}
+
+			id = fmt.Sprintf("cluster-%s", clusterName)
+		} else {
+			id = fmt.Sprintf("machine-%s", nodeID)
```

Comment:
```suggestion
			id ="machine-" + nodeID
```

Minor nit: perhaps this is clearer?